File: src/rnet/packets/ZoneParameterPacket.js
const DataPacket = require("./DataPacket");
const ExtraZoneParam = require("../extraZoneParam");

class ZoneParameterPacket extends DataPacket {
    requiresHandshake() {
        return true;
    }

    getControllerID() {
        return this.sourceControllerID;
    }

    getZoneID() {
        return this.sourcePath[2];
    }

    getParameterID() {
        return this.sourcePath[4];
    }

    getValue() {
        const byte = this.data.readUInt8(0);

        switch (parameterID) {
            case ExtraZoneParam.BASS:
            case ExtraZoneParam.TREBLE:
            case ExtraZoneParam.BALANCE:
                return byte - 10;
            case ExtraZoneParam.TURN_ON_VOLUME:
                return byte * 2;
            case ExtraZoneParam.BACKGROUND_COLOR:
            case ExtraZoneParam.PARTY_MODE:
                return byte;
            default:
                return byte == 0x01;
        }
    }
}

ZoneParameterPacket.fromPacket = function(dataPacket) {
    if (dataPacket instanceof DataPacket) {
        const zoneParameterPacket = new ZoneParameterPacket();
        dataPacket.copyToPacket(zoneParameterPacket);
        return zoneParameterPacket;
    }
    else {
        throw new TypeError("Cannot create ZoneParameterPacket with anything other than a DataPacket");
    }
}

module.exports = ZoneParameterPacket;

File: src/RNetPi.Core/RNet/ZoneParameterPacket.cs
using System;

namespace RNetPi.Core.RNet;

/// <summary>
/// Represents an RNet Zone Parameter packet
/// </summary>
public class ZoneParameterPacket : DataPacket
{
    public ZoneParameterPacket()
    {
        MessageType = 0x00;
    }

    public byte GetControllerID()
    {
        return SourceControllerID;
    }

    public byte GetZoneID()
    {
        return SourcePath.Length > 2 ? SourcePath[2] : (byte)0;
    }

    public byte GetParameterID()
    {
        return SourcePath.Length > 4 ? SourcePath[4] : (byte)0;
    }

    public byte GetParameterValue()
    {
        return Data.Length > 0 ? Data[0] : (byte)0;
    }

    /// <summary>
    /// Creates a ZoneParameterPacket from a DataPacket
    /// </summary>
    public static ZoneParameterPacket FromPacket(DataPacket dataPacket)
    {
        if (dataPacket.MessageType != 0x00)
        {
            throw new ArgumentException("Cannot create ZoneParameterPacket from packet with MessageType != 0x00");
        }

        var zoneParameterPacket = new ZoneParameterPacket();
        dataPacket.CopyToPacket(zoneParameterPacket);
        return zoneParameterPacket;
    }
}